src/NServiceBus.Core/Reliability/SynchronizedStorage/CompletableSynchronizedStorageSession.cs

```diff
@@ -0,0 +1,46 @@
+﻿namespace NServiceBus.Persistence
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Extensibility;
+    using Outbox;
+    using Transport;
+
+    /// <summary>
+    /// Represents a storage session from point of view of the infrastructure.
+    /// </summary>
+    public interface ICompletableSynchronizedStorageSession : ISynchronizedStorageSession, IDisposable
+    {
+        /// <summary>
+        /// Begins a new storage session which is an atomic unit of work.
+        /// </summary>
+        /// <param name="transaction">Outbox transaction.</param>
+        /// <param name="context">Context.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> to observe.</param>
+        /// <returns><c>true</c> when the session was opened; otherwise <c>false</c>.</returns>
+        ValueTask<bool> OpenSession(IOutboxTransaction transaction, ContextBag context, CancellationToken cancellationToken = default);
```

Comment:
Should we consider `TryOpenSession` as the name?

Could we have a single `OpenSession` method instead of 3?

Reply:
Szymon wanted to have the original adapter design preserved because it makes the expectations of the core clear. I have introduced an internal extension method that does it. Will push soon. We can make it public if we want to